File: 0212-word-search-ii/0212-word-search-ii.java
class Solution {
    class TrieNode {
        TrieNode[] children = new TrieNode[26];
        String word;
}

    public TrieNode constructer(String[] words) {
        TrieNode root = new TrieNode();
        for(String str : words) {
            TrieNode Node = root;
            for(char c : str.toCharArray()) {
                int ind = c - 'a';
                if(Node.children[ind] == null) {
                    Node.children[ind] = new TrieNode();
                }
                Node = Node.children[ind];
            }
            Node.word = str;
        }

        return root;
    }

    public List<String> findWords(char[][] board, String[] words) {
        int n = board.length, m = board[0].length;

        List<String> res = new ArrayList<>();
        TrieNode root = constructer(words);

        for(int i = 0; i < n; i++) {
            for(int j = 0; j < m; j++) {
                dfs(board, i, j, root, res);
            }
        }

        return res;
    }

    public void dfs(char[][] board, int x, int y, TrieNode root, List<String> res) {
        int n = board.length, m = board[0].length;
        if(x < 0 || x >= n || y < 0 || y >= m || board[x][y] == '*' || root.children[board[x][y] - 'a'] == null) {
            return;
        }
        char c = board[x][y];
        root = root.children[c - 'a'];

        if(root.word != null) {
            res.add(root.word);
            root.word = null;
        }
        board[x][y] = '*';
        dfs(board, x+1, y, root, res);
        dfs(board, x-1, y, root, res);
        dfs(board, x, y+1, root, res);
        dfs(board, x, y-1, root, res);
        board[x][y] = c;
    }
}

File: 0212-word-search-ii/0212-word-search-ii.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.endOfWord = False
        self.word = None  # \U0001f195 store full word here


class Solution(object):
    def findWords(self, board, words):
        m = len(board)
        n = len(board[0])
        res = set()
        self.root = TrieNode()

        def dfs(r, c, visited, curr):
            char = board[r][c]
            if char not in curr.children:
                return

            visited[r][c] = True
            curr = curr.children[char]

            if curr.endOfWord:
                res.add(curr.word)
                curr.endOfWord = False  # avoid duplicates

            for dx, dy in [[0,1], [1,0], [-1,0], [0,-1]]:
                nr = r + dx
                nc = c + dy

                if 0 <= nr < m and 0 <= nc < n and not visited[nr][nc]:
                    dfs(nr, nc, visited, curr)

            visited[r][c] = False  # \U0001f501 backtrack

        # Build the Trie
        for word in words:
            curr = self.root
            for c in word:
                if c not in curr.children:
                    curr.children[c] = TrieNode()
                curr = curr.children[c]
            curr.endOfWord = True
            curr.word = word  # \U0001f195 store the word directly

        # Start DFS
        for i in range(m):
            for j in range(n):
                if board[i][j] in self.root.children:
                    visited = [[False for _ in range(n)] for _ in range(m)]
                    dfs(i, j, visited, self.root)

        return list(res)
